File: rhizome/replication/replicator.py
"""
Система репликации данных на основе популярности
"""

import asyncio
from typing import Dict, List, Optional

from rhizome.dht.protocol import DHTProtocol
from rhizome.logger import get_logger
from rhizome.popularity.ranking import RankedItem
from rhizome.storage.storage import Storage


class Replicator:
    """Репликатор данных на основе популярности"""

    def __init__(
        self,
        dht_protocol: DHTProtocol,
        storage: Storage,
        min_replication_factor: int = 5,
        popular_replication_factor: int = 10,
    ):
        self.dht_protocol = dht_protocol
        self.storage = storage
        self.min_replication_factor = min_replication_factor
        self.popular_replication_factor = popular_replication_factor
        self.logger = get_logger("replication.replicator")

    async def replicate_popular_items(
        self, ranked_items: List[RankedItem], popularity_threshold: float = 7.0
    ) -> Dict[bytes, bool]:
        """
        Репликация популярных элементов

        Args:
            ranked_items: Отсортированный список элементов по популярности
            popularity_threshold: Порог популярности для репликации

        Returns:
            Словарь результатов репликации (key -> success)
        """
        results = {}

        # Фильтруем популярные элементы
        popular_items = [item for item in ranked_items if item.score >= popularity_threshold]

        self.logger.info(
            "Starting replication", total_items=len(ranked_items), popular_items=len(popular_items)
        )

        # Реплицируем каждый популярный элемент
        for item in popular_items:
            key = item.key
            try:
                # Получаем значение из storage
                value = await self.storage.get(key)
                if value is None:
                    self.logger.warning("Value not found for replication", key=key.hex()[:16])
                    results[key] = False
                    continue

                # Проверяем текущий replication factor
                current_replication = item.metrics.replication_count
                target_replication = self.popular_replication_factor

                if current_replication >= target_replication:
                    # Уже достаточно репликаций
                    results[key] = True
                    continue

                # Реплицируем через STORE
                # Используем увеличенный TTL для популярных данных
                ttl = 2592000  # 30 дней
                success = await self.dht_protocol.store(key, value, ttl)

                results[key] = success

                if success:
                    self.logger.debug(
                        "Replicated popular item",
                        key=key.hex()[:16],
                        score=item.score,
                        target_replication=target_replication,
                    )
                else:
                    self.logger.warning("Replication failed", key=key.hex()[:16])

            except Exception as e:
                self.logger.error(
                    "Error replicating item", key=key.hex()[:16], error=str(e), exc_info=True
                )
                results[key] = False

        successful = sum(1 for v in results.values() if v)
        self.logger.info(
            "Replication completed",
            total=len(results),
            successful=successful,
            failed=len(results) - successful,
        )

        return results

    async def ensure_minimal_replication(
        self, keys: List[bytes], min_factor: Optional[int] = None
    ) -> Dict[bytes, bool]:
        """
        Обеспечение минимальной репликации для списка ключей

        Args:
            keys: Список ключей для проверки
            min_factor: Минимальный фактор репликации (по умолчанию из конфига)

        Returns:
            Словарь результатов (key -> success)
        """
        if min_factor is None:
            min_factor = self.min_replication_factor

        results = {}

        for key in keys:
            try:
                # Получаем значение
                value = await self.storage.get(key)
                if value is None:
                    results[key] = False
                    continue

                # TODO: Проверить реальный replication factor через DHT
                # Пока просто реплицируем
                success = await self.dht_protocol.store(key, value, ttl=86400)
                results[key] = success

            except Exception as e:
                self.logger.error("Error ensuring replication", key=key.hex()[:16], error=str(e))
                results[key] = False

        return results

    async def emergency_replication(self, key: bytes, value: bytes) -> bool:
        """
        Экстренная репликация при обнаружении потери узла

        Args:
            key: Ключ для репликации
            value: Значение для репликации

        Returns:
            True если успешно реплицировано
        """
        self.logger.warning("Emergency replication", key=key.hex()[:16])

        try:
            # Реплицируем с высоким приоритетом
            # Используем увеличенный TTL
            success = await self.dht_protocol.store(key, value, ttl=2592000)  # 30 дней

            if success:
                self.logger.info("Emergency replication successful", key=key.hex()[:16])
            else:
                self.logger.error("Emergency replication failed", key=key.hex()[:16])

            return success

        except Exception as e:
            self.logger.error(
                "Error in emergency replication", key=key.hex()[:16], error=str(e), exc_info=True
            )
            return False

File: src/replication/replicator.rs
use std::collections::HashMap;
use std::sync::Arc;
use tracing::{debug, error, info, warn};

use crate::dht::protocol::DHTProtocol;
use crate::popularity::ranking::RankedItem;
use crate::storage::main::Storage;

/// Duplicate data to the other node
pub struct Replicator {
    /// DHT protocol structure
    dht_protocol: Arc<DHTProtocol>,
    /// Access to the local storage with our node data
    storage: Arc<Storage>,
    /// How many replications should this data has
    min_replication_factor: usize,
    /// How many replications should be if data very popular
    popular_replication_factor: usize,
}

impl Replicator {
    pub fn new(
        dht_protocol: Arc<DHTProtocol>,
        storage: Arc<Storage>,
        min_replication_factor: usize,
        popular_replication_factor: usize,
    ) -> Self {
        Self {
            dht_protocol,
            storage,
            min_replication_factor,
            popular_replication_factor,
        }
    }

    /// Replication of popular elements
    ///
    /// Work smth like CDN network
    pub async fn replicate_popular_items(
        &self,
        ranked_items: Vec<RankedItem>,
        popularity_threshold: f64,
    ) -> HashMap<Vec<u8>, bool> {
        let mut results = HashMap::new();

        let popular_items: Vec<&RankedItem> = ranked_items
            .iter()
            .filter(|item| item.score >= popularity_threshold)
            .collect();

        info!(
            total_items = ranked_items.len(),
            popular_items = popular_items.len(),
            "Starting replication"
        );

        for item in popular_items {
            let key = &item.key;
            let key_hex = hex::encode(&key[..key.len().min(8)]);

            let value_result = self.storage.get(key.clone()).await;

            match value_result {
                Ok(Some(value)) => {
                    let current_replication = item.metrics.replication_count as usize;
                    let target_replication = self.popular_replication_factor;

                    if current_replication >= target_replication {
                        results.insert(key.clone(), true);
                        continue;
                    }

                    let ttl = 2592000;
                    match self.dht_protocol.store(key, &value, ttl).await {
                        Ok(success) => {
                            results.insert(key.clone(), success);
                            if success {
                                debug!(
                                    key = %key_hex,
                                    score = item.score,
                                    target_replication = target_replication,
                                    "Replicated popular item"
                                );
                            } else {
                                warn!(key = %key_hex, "Replication failed");
                            }
                        }
                        Err(e) => {
                            error!(key = %key_hex, error = %e, "Error during STORE in replication");
                            results.insert(key.clone(), false);
                        }
                    }
                }
                Ok(None) => {
                    warn!(key = %key_hex, "Value not found for replication");
                    results.insert(key.clone(), false);
                }
                Err(e) => {
                    error!(key = %key_hex, error = %e, "Error accessing storage for replication");
                    results.insert(key.clone(), false);
                }
            }
        }

        let successful = results.values().filter(|&&v| v).count();
        info!(
            total = results.len(),
            successful = successful,
            failed = results.len() - successful,
            "Replication completed"
        );

        results
    }

    /// Replication for basic data
    ///
    /// Algo only send this data once to every node in network for their minimal life
    pub async fn ensure_minimal_replication(
        &self,
        keys: Vec<Vec<u8>>,
        min_factor: Option<usize>,
    ) -> HashMap<Vec<u8>, bool> {
        let _target_factor = min_factor.unwrap_or(self.min_replication_factor);
        let mut results = HashMap::new();

        for key in keys {
            let _key_hex = hex::encode(&key[..key.len().min(8)]);

            match self.storage.get(key.clone()).await {
                Ok(Some(value)) => {
                    // Выполняем STORE для обеспечения наличия данных (TTL 1 день)
                    match self.dht_protocol.store(&key, &value, 86400).await {
                        Ok(success) => results.insert(key, success),
                        Err(_) => results.insert(key, false),
                    };
                }
                _ => {
                    results.insert(key, false);
                }
            }
        }

        results
    }

    /// Panic replication
    ///
    /// If node leave us bad data should be sent for do not die
    pub async fn emergency_replication(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
        let key_hex = hex::encode(&key[..key.len().min(8)]);
        warn!(key = %key_hex, "Emergency replication triggered");

        let ttl = 2592000;
        match self.dht_protocol.store(&key, &value, ttl).await {
            Ok(true) => {
                info!(key = %key_hex, "Emergency replication successful");
                true
            }
            Ok(false) => {
                error!(key = %key_hex, "Emergency replication failed (store returned false)");
                false
            }
            Err(e) => {
                error!(key = %key_hex, error = %e, "Error in emergency replication");
                false
            }
        }
    }
}
